### ISSUE TITLE
Migration: replicating the request body of legacy webhook actions

### ISSUE THREAD

**Issue by rowanseymour** [nyaruka/goflow#70]
These currently send our own defined format of run steps, whereas the new call_webhook action has a body template.

**Comment by nicpottier** [comment]
Ya.. there's a bit of work required to allow us to fully represent our current format using a template, but also a good test that we are exposing everything necessary for other callers so worth it.

I think the biggest thing is figuring out how to represent say the path like we do in the old world and what that would look like. Maybe we need `filter` and `map` in expressions?

**Comment by rowanseymour** [comment]
What makes it extra challenging is that much of the data doesn't exist in goflow. The payload we currently send looks like:

```python
dict(channel=channel.id if channel else -1,
     channel_uuid=channel.uuid if channel else None,
     relayer=channel.id if channel else -1,
     flow=flow.id,
     flow_uuid=flow.uuid,
     flow_name=flow.name,
     flow_base_language=flow.base_language,
     run=run.id,
     text=text,
     attachments=[a.url for a in attachments],
     step=six.text_type(node_uuid),
     phone=contact.get_urn_display(org=org, scheme=TEL_SCHEME, formatted=False),
     contact=contact.uuid,
     contact_name=contact.name,
     urn=six.text_type(contact_urn),
     values=json.dumps(values),
     steps=json.dumps(steps),
     time=json_time,
     header=header)
```
where each step looks like:

```python
dict(type=step.step_type,
     node=step.step_uuid,
     arrived_on=datetime_to_str(step.arrived_on),
     left_on=datetime_to_str(step.left_on),
     text=step.get_text(),
     value=step.rule_value)
```

It's sent as urlencoded (?) form data tho steps and values are JSON encoded within that.

I'm leaning toward we introduce v2 webhooks now that have customizable payloads with only context items that we can migrate.

**Comment by rowanseymour** [comment]
Actually come to think of it, v2 webhooks in old world don't need to have customizable payloads, they just need to have payloads we know we can migrate in the new world, i.e. no database ids, no step types

**Comment by nicpottier** [comment]
What do you mean by it not existing in goflow? Don't all those fields exist (even if not in the context currently) in the new world?

**Comment by rowanseymour** [comment]
We don't send database id fields or have a concept of step_types. I'm all for adding cool mapping functions to excellent but I'm not completely convinced we can replicate this madness in an expression.

**Comment by nicpottier** [comment]
Ya, step type especially.. but how do we move everyone off then? Just settle on a new payload and send out a notification that this will be changed on x date?

**Comment by rowanseymour** [comment]
I think so tho we'd have to support both payloads for a while to give people a chance to migrate. Maybe have a v1/v2 dropdown on the action in the editor?

Arguably we should have done this a while ago since the current payload has accumulated quite a bit of cruft.

Question is do we add support for a completely customizable request body like in the new engine, or settle on a fixed schema that we can replicate as a customizable request body in the new engine. Old engine expressions context doesn't expose as much as new engine, and old engine functions definitely don't support rendering lists of steps and values etc.

It would be nice if webhook actions didn't have to send the complete run history every time - I can't imagine most people are using that.

**Comment by nicpottier** [comment]
Seems easiest for users would be to pick something similar to the current thing but what we could support in the new engine, something that we think for most users would require no action. And just make that the new "default" in the old world.

Ideally that would be the same default as the new world but I don't think it has to be since we can just populate it appropriately.

**Comment by rowanseymour** [comment]
See https://github.com/nyaruka/goflow/blob/master/flows/definition/legacy.go#L34. Still have some challenges if we want to match the old engine format exactly when migrating old flows.

**Comment by rowanseymour** [comment]
Fixed in #263
